### PR TITLE
Contribute to exclude property from validation

### LINKED ISSUE #95: Contribute to exclude property from validation
See the issue https://github.com/redhat-developer/vscode-microprofile/issues/22

Today user can exclude properties to ignore validation with settings. This issue gives the capability to contribute for excluded properties with Java SPI.

//cc @apupier

### PR DESCRIPTION
**Pull request by angelozerr**
Contribute to exclude property from validation

Fixes #95

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
I will take a look at this today

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/test/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettingsTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/*******************************************************************************
+* Copyright (c) 2019 Red Hat Inc. and others.
```
Comment (datho7561) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/test/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettingsTest.java:2]: 2020

File: microprofile.ls/README.md [hunk 2]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
 if there is the declaration `mp.messaging.outgoing.generated-price.connector=smallrye-kafka` 
 
 The `mp.messaging.outgoing.generated-price.topic` property cannot be computed on the external component side because it depends on the value of 
-`mp.messaging.outgoing.generated-price.connector`. The comput	tion is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
+`mp.messaging.outgoing.generated-price.connector`. The computation is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
 
-Please see the [sample of MicroProfile Reactive Messaging](/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
\ No newline at end of file
+Please see the [sample of MicroProfile Reactive Messaging](/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
```
Comment (datho7561) [inline microprofile.ls/README.md:139]: This link is broken (should have `org.eclipse.lsp4mp.ls` at the beginning of the path)

File: microprofile.ls/README.md [hunk 3]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
 if there is the declaration `mp.messaging.outgoing.generated-price.connector=smallrye-kafka` 
 
 The `mp.messaging.outgoing.generated-price.topic` property cannot be computed on the external component side because it depends on the value of 
-`mp.messaging.outgoing.generated-price.connector`. The comput	tion is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
+`mp.messaging.outgoing.generated-price.connector`. The computation is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
```
Comment (datho7561) [inline microprofile.ls/README.md:137]: This link is broken (should have `org.eclipse.lsp4mp.ls` at the beginning of the path)

**Review by datho7561 — CHANGES_REQUESTED** [review]
Should the `camel.*` exclusion apply by default? From my testing, the Camel properties are still validated by default.

File: microprofile.ls/README.md (hunk 2 above)
Comment (angelozerr) [reply microprofile.ls/README.md]: fixed

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/test/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettingsTest.java (hunk 1 above)
Comment (angelozerr) [reply microprofile.ls/org.eclipse.lsp4mp.ls/src/test/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettingsTest.java]: fixed

File: microprofile.ls/README.md [hunk 4]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
...
 
-Please see the [sample of MicroProfile Reactive Messaging](/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
\ No newline at end of file
+Please see the [sample of MicroProfile Reactive Messaging](/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
+
+Contribute to settings
+-------
+
+In microprofile-config.properties file when a property doesn't exists, there is an unkwown error. To ignore this error, user must update the client settings to ignore this error.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply microprofile.ls/README.md]: fixed

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettings.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
+/*******************************************************************************
+* Copyright (c) 2019 Red Hat Inc. and others.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileExtensionSettings.java]: fixed

File: microprofile.ls/README.md (hunk 3 above)
Comment (angelozerr) [reply microprofile.ls/README.md]: fixed

**Comment by angelozerr** [comment]
> Should the camel.* exclusion apply by default? From my testing, the Camel properties are still validated by default.

No it's not done by default (it's just for test). @apupier should implement in the camel extension to enable that.

**Review by datho7561 — CHANGES_REQUESTED** [review]
I tried to make [an extension that would contribute an exclusion for all `helidon.*` properties](https://github.com/datho7561/test-properties-extension), but I can't seem to get it to work. It could just be how I set up the extension though, this is the first time I've done this.

**Comment by angelozerr** [comment]
> I tried to make an extension that would contribute an exclusion for all helidon.* properties, but I can't seem to get it to work. It could just be how I set up the extension though, this is the first time I've done this.

It's MP LS extension (like quarkus.ls.ext) and not a MP JDT extension. To test quickly add this settings.json in the quarkus.ls.ext/META-INF/lsp4mp folder and rebuild it.

File: microprofile.ls/README.md [hunk 6]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
 if there is the declaration `mp.messaging.outgoing.generated-price.connector=smallrye-kafka` 
 
 The `mp.messaging.outgoing.generated-price.topic` property cannot be computed on the external component side because it depends on the value of 
-`mp.messaging.outgoing.generated-price.connector`. The comput	tion is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
+`mp.messaging.outgoing.generated-price.connector`. The computation is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply microprofile.ls/README.md]: link use the full path url now.

File: microprofile.ls/README.md [hunk 7]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
...
 
-Please see the [sample of MicroProfile Reactive Messaging](/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
\ No newline at end of file
+Please see the [sample of MicroProfile Reactive Messaging](/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply microprofile.ls/README.md]: link use the full path url now.

**Review by datho7561 — APPROVED** [review]

**Comment by datho7561** [comment]
Looks good to me!

File: microprofile.ls/README.md [hunk 8]
```diff
@@ -134,6 +134,28 @@ In other words, `mp.messaging.outgoing.generated-price.topic` exists only
...
\ No newline at end of file
+Please see the [sample of MicroProfile Reactive Messaging](https://github.com/eclipse/lsp4mp/blob/master/microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/extensions/reactivemessaging).
+
+Contribute to settings
+-------
+
+In microprofile-config.properties file when a property doesn't exists, there is an unkwown error. The user can configure which unknown property errors to hide in the client settings.
+It's possible to ignore this error with contribution (ex : camel toolings can ignore validation for all `camel.*` properties).
+
+To do that in the JAR extension, create the `META-INF/lsp4mp/settings.json` like this:
```
Comment (datho7561) [inline microprofile.ls/README.md:147]: Mention creating a lsp4mp extension, link to the documentation for creating an lsp4mp extension. Then mention that the `META-INF/lsp4mp/settings.json` should be put into the lsp4mp extension.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: microprofile.ls/README.md (hunk 8 above)
Comment (angelozerr) [reply microprofile.ls/README.md:156]: @datho7561 there is a new section `LSPMP - LS - extensions`, please tell me if it's better.

**Review by datho7561 — APPROVED** [review]

File: microprofile.ls/README.md (hunk 8 above)
Comment (datho7561) [reply microprofile.ls/README.md:156]: It looks good!

**Comment by datho7561** [comment]
There appears to be a compilation error

**Pull request merged**